jstz_api/src/urlpattern/mod.rs

```diff
@@ -0,0 +1,576 @@
+            .exec(input, base_url)?
+            .map_or(Ok(JsValue::Null), |r| Ok(r.into_js(context)))
+    }
+}
+
+impl TryFromJs for UrlPatternInit {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        let obj = value.as_object().ok_or_else(|| {
+            JsError::from_native(JsNativeError::typ().with_message("Expected `JsObject`"))
+        })?;
+
+        macro_rules! get_optional_property {
+            ($obj:ident, $field:literal, $context:ident) => {
+                if $obj.has_property(js_string!($field), $context)? {
+                    $obj.get(js_string!($field), $context)?
+                        .try_js_into($context)?
+                } else {
+                    None
+                }
+            };
+        }
+
+        let url_pattern_init = urlpattern::quirks::UrlPatternInit {
+            protocol: get_optional_property!(obj, "protocol", context),
+            username: get_optional_property!(obj, "username", context),
+            password: get_optional_property!(obj, "password", context),
+            hostname: get_optional_property!(obj, "hostname", context),
+            port: get_optional_property!(obj, "port", context),
+            pathname: get_optional_property!(obj, "pathname", context),
+            search: get_optional_property!(obj, "search", context),
+            hash: get_optional_property!(obj, "hash", context),
+            base_url: get_optional_property!(obj, "base_url", context),
+        };
+
+        Ok(Self(url_pattern_init))
+    }
+}
+
+impl TryFromJs for UrlPatternInput {
+    fn try_from_js(value: &JsValue, context: &mut Context<'_>) -> JsResult<Self> {
+        if let Some(string) = value.as_string() {
+            return Ok(Self(InnerStringOrQuirksInit::String(
+                string.to_std_string_escaped(),
+            )));
+        };
+
+        let UrlPatternInit(init) = UrlPatternInit::try_from_js(value, context)?;
+        Ok(Self(InnerStringOrQuirksInit::Init(init)))
+    }
+}
+
+impl IntoJs for UrlPatternInput {
+    fn into_js(self, context: &mut Context<'_>) -> JsValue {
+        let UrlPatternInput(string_or_init) = self;
+        match string_or_init {
+            InnerStringOrQuirksInit::Init(init) => UrlPatternInit(init).into_js(context),
+            InnerStringOrQuirksInit::String(string) => JsString::from(string).into(),
+        }
+    }
+}
+
+impl IntoJs for UrlPatternComponentResult {
+    fn into_js(self, context: &mut Context<'_>) -> JsValue {
+        let UrlPatternComponentResult(url_pattern_component_result) = self;
+        let input = url_pattern_component_result.input;
+        let groups: Vec<(String, String)> =
+            url_pattern_component_result.groups.into_iter().collect();
+        let obj = JsObject::with_object_proto(context.intrinsics());
+        let _ = obj.create_data_property(
+            JsString::from("input"),
+            JsValue::String(JsString::from(input)),
+            context,
+        );
```

Comment:
Comments here for others would be helpful (in future) -- not everyone is familiar with the way of constructing objects using `JsObject`

Reply:
Added a small comment making a reference to the prototype thing.